pkg/scheduler/plugins/gpushare/plugin.go

```diff
@@ -0,0 +1,285 @@
+/*
+Copyright 2022 The Koordinator Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gpushare
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/kubernetes/pkg/api/v1/resource"
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+
+	apiext "github.com/koordinator-sh/koordinator/apis/extension"
+	schedulingv1alpha1 "github.com/koordinator-sh/koordinator/apis/scheduling/v1alpha1"
+	"github.com/koordinator-sh/koordinator/pkg/scheduler/frameworkext"
+	"github.com/koordinator-sh/koordinator/pkg/scheduler/plugins/names"
+)
+
+const (
+	// Name is the name of the plugin used in the plugin registry and configurations.
+	Name = names.GPUShare
+
+	// stateKey is the key in CycleState to pre-computed data.
+	stateKey = Name
+
+	// ErrMissingDevice when node does not have Device.
+	ErrMissingDevice = "node(s) missing Device"
+
+	// ErrNoMatchNode when node can't satisfy Pod's requested resource.
+	ErrNoMatchNode = "node(s) didn't match the Pod resource requirement"
+)
+
+type Plugin struct {
+	handle framework.Handle
+	// deviceClient    schedulingclient.DeviceInterface
+	// deviceLister    schedulinglisters.DeviceLister
+	// nodeLister      corelisters.NodeLister
+	nodeDeviceCache *nodeDeviceCache
+}
+
+var (
+	_ framework.PreFilterPlugin = &Plugin{}
+	_ framework.FilterPlugin    = &Plugin{}
+	_ framework.ReservePlugin   = &Plugin{}
+	_ framework.PreBindPlugin   = &Plugin{}
+)
+
+type preFilterState struct {
+	skip              bool
+	allocResult       []*apiext.DeviceAllocation
+	convertedResource corev1.ResourceList
+}
+
+func (s *preFilterState) Clone() framework.StateData {
+	return s
+}
+
+func (g *Plugin) Name() string {
+	return Name
+}
+
+func (g *Plugin) PreFilter(ctx context.Context, cycleState *framework.CycleState, pod *corev1.Pod) *framework.Status {
+	state := &preFilterState{skip: true}
+
+	podRequest, _ := resource.PodRequestsAndLimits(pod)
+
+	if isGPUPod(podRequest) {
+		combination, err := validateGPURequest(podRequest)
```

Comment:
The plugin in actually only handle GPU case. actually in prefilter, you should classifiy resources by device, and "for device...". If GPU is special, try use if device == GPU to handle. The diffculty is that how to  classifiy resources by device, you can mantain a device-resource-map in device_handler\cache,  when device crd udpated, you can update the device-resource-map. this is Automatically for all devices new-added, but you should pay attention that you should first load all device crds to rebuild the device-resource-map then load pod, may be you need execute WaitForCacheSync to make it happen. Or you may maintain an artificial rule configuration, the bad case is that you have to update the  rule configuration if new device added.

Reply:
That's a really good suggestion. I'll leave a TODO here, and try to handle all kind of devices in the next few PRs.